src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,754 @@
+            | _ -> t (*Not supported right now*)
+          else if var_count = 2 then
+            let get_vars i a l = if Z.equal a Z.zero then l else (i, a)::l in
+            let v12 = Array.fold_righti get_vars expr [] in
+            let a1 = snd (List.hd v12) in
+            let a2 = snd (List.hd @@ List.tl v12) in
+            let var1 = fst (List.hd v12) in
+            let var2 = fst (List.hd @@ List.tl v12) in
+            match Tcons1.get_typ tcons with
+            | EQ ->
+              let res =
+                if Z.equal a1 Z.one && Z.equal a2 Z.(-one)
+                then meet_with_one_conj t var2 (Some var1, !constant)
+                else if Z.equal a1 Z.(-one) && Z.equal a2 Z.one
+                then meet_with_one_conj t var1 (Some var2, !constant)
+                else t
+              in overflow_handling res original_expr
+            | _-> t (*Not supported right now*)
+          else
+            t (*For any other case we don't know if the (in-) equality is true or false or even possible therefore we just return t *)
+
+  let meet_tcons t tcons expr = timing_wrap "meet_tcons" (meet_tcons t tcons) expr
+
+  let unify a b =
+    meet a b
+
+  let unify a b =
+    let res = unify a b  in
+    if M.tracing then M.tracel "ops" "unify: %s %s -> %s\n" (show a) (show b) (show res);
+    res
+
+  (* Assert a constraint expression. Defined in apronDomain.apron.ml
+
+     If the constraint is never fulfilled, then return bottom.
+     Else the domain can be modified with the new information given by the constraint.
+
+     It basically just calls the function meet_tcons.
+
+     It is called by eval (defined in sharedFunctions), but also when a guard in
+     e.g. an if statement is encountered in the C code.
+
+  *)
+  let assert_cons d e negate no_ov =
+    let no_ov = Lazy.force no_ov in
+    if M.tracing then M.tracel "assert_cons" "assert_cons with expr: %a %b\n" d_exp e no_ov;
+    match Convert.tcons1_of_cil_exp d d.env e negate no_ov with
+    | tcons1 -> meet_tcons d tcons1 e
+    | exception Convert.Unsupported_CilExp _ -> d
+
+  let assert_cons d e negate no_ov = timing_wrap "assert_cons" (assert_cons d e negate) no_ov
+
+  let relift t = t
+
+  (* representation as C expression
+
+     This function returns all the equalities that are saved in our datastructure t.
+
+     Lincons -> linear constraint *)
+  let invariant t =
+    match t.d with
+    | None -> []
+    | Some d ->
+      let earray = Lincons1.array_make t.env (Array.length d) in
+      for i = 0 to Array.length d - 1 do
+        let (var_opt, const) = d.(i) in
+        let coeff_vars = match var_opt with
+          | None -> []
+          | Some var_index ->
+            let var = Environment.var_of_dim t.env var_index in
+            [(Coeff.s_of_int 1, var)]
+        in
+        let cst = Coeff.s_of_int (Z.to_int const) in
+        Lincons1.set_list (Lincons1.array_get earray i) coeff_vars (Some cst)
+      done;
```

Comment:
I am rather sceptical, that this part really represents the invariants that the analysis has found. In particular, linear constraints are represented in apron in a normalized fashion, i.e. with all terms on one side of the constraint, and the other side being constantly zero.

Thus, the purpose of the coeff_vars would be the provider of all terms, which would usually include some term, whose index corresponds to its position in the original d array. I can not see such a term being created within function invariant.

Furthermore, I am wondering, whether one could simply create a List of Lincons1.t directly without the pesky detour via Lincons1 array.

Also: the for loop can be easily replaced with an Array.iteri

Reply:
I think it would be best to try this out: Passing the additional option `--enable witness.yaml.enabled ` to Goblint will produce a YAML file providing invariants for program points which you can then inspect to see if the invariants are printed correctly.